Userland/DevTools/HackStudio/ProjectTemplateManager.h

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2021, Nick Vella <nick@nxk.io>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "ProjectTemplate.h"
+#include <AK/NonnullRefPtrVector.h>
+
+namespace HackStudio {
+
+class ProjectTemplateManager {
+public:
+    ProjectTemplateManager();
+
+    static ProjectTemplateManager& the();
+
+    void rescan_templates();
```

Comment:
If I'm not mistaken, this is only called once (when creating the `ProjectTemplateManager` singleton), and then never again until HackStudio is closed.

Ideally, it would be nice to watch both the `/res/devel/templates/` directory for additions/deletions, and each identified `.ini` file into it for changes. However, after looking into it, these kind of utilities aren't really available in `Core` :
* The DirectoryWatcher is blocking and probably wouldn't play nice with the EventLoop. It seems like it was more created for [CrashDaemon](https://github.com/SerenityOS/serenity/blob/master/Userland/Services/CrashDaemon/main.cpp#L92) than for general purpose, and will need to be adapted for this use-case.
* There isn't any good wrapper for watching a file. The `watch_file` syscall is manually used for the GUI in [FileSystemModel.cpp line 142](https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibGUI/FileSystemModel.cpp#L142) via a Notifier, but not wrapped in any Core class.

Creating these utilities look very out of scope for this PR, though, and shouldn't block it on that.

It would however be nice to at least have one "refresh templates" button somewhere in the template UI (or maybe in a menu).

Reply:
Consolidated the `ProjectTemplateManager` into the `ProjectTemplatesModel` for now; after a bit of a think I've deduced it to really only be added complexity, all template creation logic is in `ProjectTemplate` anyway. This makes using the new `FileWatcher` a lot easier, which I've gone ahead and done :)